sync/sync-impl/src/main/java/com/duckduckgo/sync/impl/SyncAccountRepository.kt

```diff
@@ -95,19 +114,139 @@ class AppSyncAccountRepository @Inject constructor(
+                Adapters.recoveryCodeAdapter.fromJson(decodedCode)?.connect != null -> CodeType.CONNECT
+                Adapters.invitationCodeAdapter.fromJson(decodedCode)?.exchangeKey != null -> CodeType.EXCHANGE
+                else -> UNKNOWN
+            }
+        }.onFailure {
+            Timber.e(it, "Failed to decode code")
+        }.getOrDefault(UNKNOWN)
+    }
+
+    private fun completeExchange(invitationCode: InvitationCode): Result<Boolean> {
+        // Sync: InviteFlow - B (https://app.asana.com/0/72649045549333/1209571867429615)
+        Timber.d("Sync-exchange: InviteFlow - B. code is an exchange code $invitationCode")
+
+        // generate new ID and and public/private key-pair for receiving device
+        val thisDeviceKeyId = deviceKeyGenerator.generate()
+        val thisDeviceKeyPair = nativeLib.prepareForConnect()
+
+        pendingInvitationAsReceiver = PendingInvitation(
+            keyId = thisDeviceKeyId,
+            privateKey = thisDeviceKeyPair.secretKey,
+            publicKey = thisDeviceKeyPair.publicKey,
+        )
+        val deviceName = syncDeviceIds.deviceName()
+
+        Timber.i(
+            "Sync: details for this (receiver) device:" +
+                "\n\tkey ID is $thisDeviceKeyId" +
+                "\n\tpublic key is ${thisDeviceKeyPair.publicKey}" +
+                "\n\tdevice name is $deviceName",
+        )
+
+        val invitedDeviceDetails = InvitedDeviceDetails(
+            keyId = thisDeviceKeyId,
+            publicKey = thisDeviceKeyPair.publicKey,
+            deviceName = deviceName,
+        )
+
+        kotlin.runCatching {
+            val payload = Adapters.invitedDeviceAdapter.toJson(invitedDeviceDetails)
+            val encrypted = nativeLib.seal(payload, invitationCode.publicKey)
+            return syncApi.sendEncryptedMessage(invitationCode.keyId, encrypted)
+        }.getOrElse { throwable ->
+            throwable.asErrorResult().alsoFireAccountErrorPixel()
+            return Error(code = GENERIC_ERROR.code, reason = "Exchange: Error encrypting payload")
+        }
+    }
+
+    override fun pollForRecoveryCodeAndLogin(): Result<ExchangeResult> {
+        // Sync: InviteFlow - E (https://app.asana.com/0/72649045549333/1209571867429615)
+        Timber.d("Sync-exchange: InviteFlow - E")
+
+        val pendingInvite = pendingInvitationAsReceiver
+            ?: return Error(code = CONNECT_FAILED.code, reason = "Connect: No pending invite initialized").also {
+                Timber.w("Sync-exchange: no pending invite initialized")
+            }
+
+        return when (val result = syncApi.getEncryptedMessage(pendingInvite.keyId)) {
+            is Error -> {
+                if (result.code == NOT_FOUND.code) {
+                    return Success(Pending)
+                } else if (result.code == GONE.code) {
+                    return Error(code = CONNECT_FAILED.code, reason = "Connect: keys expired").alsoFireAccountErrorPixel()
```

Comment:
Let's create a new code failed to differentiate from the actual connect. This is android internal, so feel free to introduce a new value just for the exchange.

Reply:
Good idea; added `EXCHANGE_FAILED`